File: frontend/android-app/app/src/main/java/com/codenized/planixor/domain/backup/BackupSerializer.kt
package com.codenized.planixor.domain.backup

import com.codenized.planixor.data.local.AnnualHoursConfigDao
import com.codenized.planixor.data.local.CalendarEventDao
import com.codenized.planixor.data.local.NotificationRecordDao
import com.codenized.planixor.data.local.PreferencesRepository
import com.codenized.planixor.data.local.ReminderDao
import com.codenized.planixor.data.local.ShiftDao
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.flow.first
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import javax.inject.Inject

/**
 * Serializes all local application data into a backup JSON string.
 * Reads all records (including soft-deleted) from Room DAOs and DataStore,
 * converts them to the platform-agnostic backup format, and produces a
 * complete BackupFile JSON with metadata.
 */
class BackupSerializer @Inject constructor(
    private val shiftDao: ShiftDao,
    private val reminderDao: ReminderDao,
    private val calendarEventDao: CalendarEventDao,
    private val notificationRecordDao: NotificationRecordDao,
    private val annualHoursConfigDao: AnnualHoursConfigDao,
    private val preferencesRepository: PreferencesRepository,
) {

    private val gson = GsonBuilder().serializeNulls().create()

    /**
     * Serializes all local data into a backup JSON string.
     * Includes all records (active and soft-deleted) from all entity tables.
     */
    suspend fun serialize(): String {
        val shifts = shiftDao.getAll().map { entity ->
            BackupShift(
                id = entity.id.lowercase(),
                name = entity.name,
                icon = entity.icon,
                backgroundColor = entity.backgroundColor,
                startTime = entity.startTime,
                endTime = entity.endTime,
                hoursWorked = entity.hoursWorked,
                isActive = entity.isActive,
                createdAt = entity.createdAt.toIsoString(),
                modifiedAt = entity.modifiedAt.toIsoString(),
                syncedAt = entity.syncedAt.toIsoStringOrNull(),
                isDeleted = entity.isDeleted,
            )
        }

        val reminders = reminderDao.getAll().map { entity ->
            BackupReminder(
                id = entity.id.lowercase(),
                name = entity.name,
                icon = entity.icon,
                backgroundColor = entity.backgroundColor,
                isActive = entity.isActive,
                createdAt = entity.createdAt.toIsoString(),
                modifiedAt = entity.modifiedAt.toIsoString(),
                syncedAt = entity.syncedAt.toIsoStringOrNull(),
                isDeleted = entity.isDeleted,
            )
        }

        val calendarEvents = calendarEventDao.getAll().map { entity ->
            BackupCalendarEvent(
                id = entity.id.lowercase(),
                eventType = entity.eventType,
                eventTypeId = entity.eventTypeId.lowercase(),
                startDay = entity.startDay,
                endDay = entity.endDay,
                startTime = entity.startTime,
                endTime = entity.endTime,
                totalHours = entity.totalHours,
                notes = entity.notes,
                alertOffsets = parseAlertOffsets(entity.alertOffsets),
                modifiedAt = entity.modifiedAt.toIsoString(),
                syncedAt = entity.syncedAt.toIsoStringOrNull(),
                isDeleted = entity.isDeleted,
            )
        }

        val notificationRecords = notificationRecordDao.getAll().map { entity ->
            BackupNotificationRecord(
                id = entity.id.lowercase(),
                calendarEventId = entity.calendarEventId.lowercase(),
                alertOffset = entity.alertOffset,
                triggerTime = entity.triggerTime.toIsoString(),
                isDelivered = entity.isDelivered,
                isRead = entity.isRead,
                modifiedAt = entity.modifiedAt.toIsoString(),
                syncedAt = entity.syncedAt.toIsoStringOrNull(),
                isDeleted = entity.isDeleted,
            )
        }

        val annualHoursConfig = annualHoursConfigDao.getAllIncludingDeleted().map { entity ->
            BackupAnnualHoursConfig(
                id = entity.id.lowercase(),
                year = entity.year,
                configuredHours = entity.configuredHours,
                modifiedAt = entity.modifiedAt.toIsoString(),
                syncedAt = entity.syncedAt.toIsoStringOrNull(),
                isDeleted = entity.isDeleted,
            )
        }

        val syncConfig = preferencesRepository.syncConfigFlow.first()
        val syncConfigList = if (syncConfig != null) {
            listOf(
                BackupSyncConfig(
                    serverUrl = syncConfig.serverUrl,
                    apiKey = syncConfig.apiKey,
                    username = syncConfig.username,
                    apiBasePath = syncConfig.apiBasePath,
                    syncIntervalMinutes = syncConfig.syncIntervalMinutes,
                    isPaused = syncConfig.isPaused,
                    lastSyncedAt = syncConfig.lastSyncedAt.toIsoStringOrNull(),
                )
            )
        } else {
            emptyList()
        }

        val backupFile = BackupFile(
            metadata = BackupMetadata(
                createdAt = Instant.now().atOffset(ZoneOffset.UTC)
                    .format(DateTimeFormatter.ISO_INSTANT),
                appVersion = APP_VERSION,
                platform = PLATFORM_ANDROID,
                schemaVersion = CURRENT_SCHEMA_VERSION,
            ),
            data = BackupData(
                calendarEvents = calendarEvents,
                notificationRecords = notificationRecords,
                annualHoursConfig = annualHoursConfig,
                shifts = shifts,
                reminders = reminders,
                syncConfig = syncConfigList,
            ),
        )

        return gson.toJson(backupFile)
    }

    private fun parseAlertOffsets(json: String): List<Int> {
        return try {
            val type = object : TypeToken<List<Int>>() {}.type
            gson.fromJson(json, type) ?: emptyList()
        } catch (_: Exception) {
            emptyList()
        }
    }

    companion object {
        private const val APP_VERSION = "1.0.0"
        private const val PLATFORM_ANDROID = "android"
    }
}

private fun Long.toIsoString(): String = Instant.ofEpochMilli(this)
    .atOffset(ZoneOffset.UTC)
    .format(DateTimeFormatter.ISO_INSTANT)

private fun Long?.toIsoStringOrNull(): String? = this?.toIsoString()

File: frontend/react-web/src/features/backup/services/backupSerializer.ts
import { db } from '@/data/db';

import type {
  BackupAnnualHoursConfig,
  BackupCalendarEvent,
  BackupFile,
  BackupNotificationRecord,
  BackupReminder,
  BackupShift,
  BackupSyncConfig,
} from '../models';
import { CURRENT_SCHEMA_VERSION } from '../models';

/** Placeholder app version — will be updated when proper version management is in place */
const APP_VERSION = '1.0.0';

/**
 * Converts a Date to ISO 8601 UTC string with Z suffix.
 */
const toIso = (date: Date): string => date.toISOString();

/**
 * Converts a nullable Date to ISO 8601 UTC string or null.
 */
const toIsoOrNull = (date: Date | null): string | null =>
  date ? date.toISOString() : null;

/**
 * Serializes all application data from IndexedDB into a BackupFile JSON string.
 *
 * Reads all entities (including soft-deleted records) from all Dexie tables,
 * converts Date objects to ISO 8601 UTC strings, UUIDs to lowercase, and
 * produces the full backup JSON with metadata and data sections.
 *
 * @returns JSON string representation of the complete BackupFile
 */
export const serializeBackup = async (): Promise<string> => {
  const [
    shifts,
    reminders,
    calendarEvents,
    notificationRecords,
    annualHoursConfig,
    syncConfigRecords,
  ] = await Promise.all([
    db.shifts.toArray(),
    db.reminders.toArray(),
    db.calendarEvents.toArray(),
    db.notifications.toArray(),
    db.annualHoursConfig.toArray(),
    db.syncConfig.toArray(),
  ]);

  const backupShifts: BackupShift[] = shifts.map((shift) => ({
    id: shift.id.toLowerCase(),
    name: shift.name,
    icon: shift.icon,
    backgroundColor: shift.backgroundColor,
    startTime: shift.startTime,
    endTime: shift.endTime,
    hoursWorked: shift.hoursWorked,
    isActive: shift.isActive,
    createdAt: toIso(shift.createdAt),
    modifiedAt: toIso(shift.modifiedAt),
    syncedAt: toIsoOrNull(shift.syncedAt),
    isDeleted: shift.isDeleted,
  }));

  const backupReminders: BackupReminder[] = reminders.map((reminder) => ({
    id: reminder.id.toLowerCase(),
    name: reminder.name,
    icon: reminder.icon,
    backgroundColor: reminder.backgroundColor,
    isActive: reminder.isActive,
    seriesFrequency: reminder.seriesFrequency,
    seriesEndDate: reminder.seriesEndDate ?? null,
    createdAt: toIso(reminder.createdAt),
    modifiedAt: toIso(reminder.modifiedAt),
    syncedAt: toIsoOrNull(reminder.syncedAt),
    isDeleted: reminder.isDeleted,
  }));

  const backupCalendarEvents: BackupCalendarEvent[] = calendarEvents.map(
    (event) => ({
      id: event.id.toLowerCase(),
      eventType: event.eventType,
      eventTypeId: event.eventTypeId.toLowerCase(),
      startDay: event.startDay,
      endDay: event.endDay,
      startTime: event.startTime,
      endTime: event.endTime,
      totalHours: event.totalHours,
      notes: event.notes ?? null,
      alertOffsets: event.alertOffsets,
      seriesId: event.seriesId ?? null,
      modifiedAt: toIso(event.modifiedAt),
      syncedAt: toIsoOrNull(event.syncedAt),
      isDeleted: event.isDeleted,
    }),
  );

  const backupNotificationRecords: BackupNotificationRecord[] =
    notificationRecords.map((record) => ({
      id: record.id.toLowerCase(),
      calendarEventId: record.calendarEventId.toLowerCase(),
      alertOffset: record.alertOffset,
      triggerTime: toIso(record.triggerTime),
      isDelivered: record.isDelivered,
      isRead: record.isRead,
      modifiedAt: toIso(record.modifiedAt),
      syncedAt: toIsoOrNull(record.syncedAt),
      isDeleted: record.isDeleted,
    }));

  const backupAnnualHoursConfig: BackupAnnualHoursConfig[] =
    annualHoursConfig.map((config) => ({
      id: config.id.toLowerCase(),
      year: config.year,
      configuredHours: config.configuredHours,
      modifiedAt: toIso(config.modifiedAt),
      syncedAt: toIsoOrNull(config.syncedAt),
      isDeleted: config.isDeleted,
    }));

  const backupSyncConfig: BackupSyncConfig[] = syncConfigRecords.map(
    (config) => ({
      serverUrl: config.serverUrl,
      apiKey: config.apiKey,
      username: config.username,
      apiBasePath: config.apiBasePath,
      syncIntervalMinutes: config.syncIntervalMinutes,
      isPaused: config.isPaused,
      lastSyncedAt: config.lastSyncedAt ?? null,
    }),
  );

  const backupFile: BackupFile = {
    metadata: {
      createdAt: new Date().toISOString(),
      appVersion: APP_VERSION,
      platform: 'web',
      schemaVersion: CURRENT_SCHEMA_VERSION,
    },
    data: {
      calendarEvents: backupCalendarEvents,
      notificationRecords: backupNotificationRecords,
      annualHoursConfig: backupAnnualHoursConfig,
      shifts: backupShifts,
      reminders: backupReminders,
      syncConfig: backupSyncConfig,
    },
  };

  return JSON.stringify(backupFile);
};
